File: docs/palindromes-in-peg/gs_flag_trace.py
"""External loading/work traces for testing the finite mirror-flag worker."""
from gs_flag_heads import compile_flags, FlagVM
from gs_dual_flags import compile_dual_flags, DualFlagVM


def trace(word, lower, program=None, dual=False):
  if set(word) - set("ab") or not 0 <= lower <= len(word):
    raise ValueError("binary word and valid lower endpoint required")
  program = (compile_dual_flags() if dual else compile_flags()) if program is None else program
  observer = (DualFlagVM if dual else FlagVM)(word, lower, len(word), program)
  observer.run()
  parts = []
  for index, char in enumerate(word):
    if index == lower:
      parts.append("|")
    parts.append(char + "." * 4)
  if lower == len(word):
    parts.append("|")
  # One GS instruction plus at most three queue work units, then halt.
  parts.extend(("!", "." * (4 * observer.steps + 1)))
  return "".join(parts)

File: scala/pal/src/main/scala/pal/GsFlagTrace.scala
package pal

/** External loading/work traces for testing the finite mirror-flag worker.
  *
  * Port of `gs_flag_trace.py`.
  */
object GsFlagTrace {

  /** The loading trace of `word` with the interval marker `|` before position
    * `lower`, then `!` and one work unit per physical transition of the flag
    * worker (`4 * steps + 1`).
    */
  def trace(word: String, lower: Int, program: Option[Program] = None, dual: Boolean = false): String = {
    if (word.exists(char => char != 'a' && char != 'b') || !(0 <= lower && lower <= word.length)) {
      throw new IllegalArgumentException("binary word and valid lower endpoint required")
    }
    val steps = if (dual) {
      val observer = new DualFlagVM(word, lower, word.length, program.orElse(Some(GsDualFlags.compileDualFlags())))
      observer.run()
      observer.steps
    } else {
      val observer = new FlagVM(word, lower, word.length, program.orElse(Some(GsFlagHeads.compileFlags())))
      observer.run()
      observer.steps
    }
    val parts = new StringBuilder
    word.zipWithIndex.foreach { case (char, index) =>
      if (index == lower) {
        parts += '|'
      }
      parts += char
      parts ++= "." * 4
    }
    if (lower == word.length) {
      parts += '|'
    }
    // One GS instruction plus at most three queue work units, then halt.
    parts += '!'
    parts ++= "." * (4 * steps + 1)
    parts.result()
  }
}
